object/host_certificate_info.go

```diff
@@ -37,42 +39,77 @@ type HostCertificateInfo struct {
 	ThumbprintSHA1   string
 	ThumbprintSHA256 string
 
+	Err         error
+	Certificate *x509.Certificate `json:"-"`
+
 	subjectName *pkix.Name
 	issuerName  *pkix.Name
 }
 
 // FromCertificate converts x509.Certificate to HostCertificateInfo
 func (info *HostCertificateInfo) FromCertificate(cert *x509.Certificate) *HostCertificateInfo {
+	info.Certificate = cert
 	info.subjectName = &cert.Subject
 	info.issuerName = &cert.Issuer
 
 	info.Issuer = info.fromName(info.issuerName)
 	info.NotBefore = &cert.NotBefore
 	info.NotAfter = &cert.NotAfter
 	info.Subject = info.fromName(info.subjectName)
-	info.Status = string(types.HostCertificateManagerCertificateInfoCertificateStatusUnknown)
 
-	{
-		sum := sha1.Sum(cert.Raw)
-		hex := make([]string, len(sum))
-		for i, b := range sum {
-			hex[i] = fmt.Sprintf("%02X", b)
-		}
-		info.ThumbprintSHA1 = strings.Join(hex, ":")
+	info.ThumbprintSHA1 = soap.ThumbprintSHA1(cert)
+
+	// SHA-256 for info purposes only, API fields all use SHA-1
+	sum := sha256.Sum256(cert.Raw)
+	hex := make([]string, len(sum))
+	for i, b := range sum {
+		hex[i] = fmt.Sprintf("%02X", b)
 	}
+	info.ThumbprintSHA256 = strings.Join(hex, ":")
 
-	{
-		sum := sha256.Sum256(cert.Raw)
-		hex := make([]string, len(sum))
-		for i, b := range sum {
-			hex[i] = fmt.Sprintf("%02X", b)
-		}
-		info.ThumbprintSHA256 = strings.Join(hex, ":")
+	if info.Status == "" {
+		info.Status = string(types.HostCertificateManagerCertificateInfoCertificateStatusUnknown)
 	}
 
 	return info
 }
 
+// FromURL connects to the given URL.Host via tls.Dial with the given tls.Config and populates the HostCertificateInfo
+// via tls.ConnectionState.  If the certificate was verified with the given tls.Config, the Err field will be nil.
+// Otherwise, Err will be set to the x509.UnknownAuthorityError or x509.HostnameError.
+// If tls.Dial returns an error of any other type, that error is returned.
+func (info *HostCertificateInfo) FromURL(u *url.URL, config *tls.Config) error {
+	addr := u.Host
+	if !(strings.LastIndex(addr, ":") > strings.LastIndex(addr, "]")) {
```

Comment:
golang stdlib uses the same idiom, but not in anything exported.  splitHostPort is in the soap package, didn't want to export it just yet as this is the only place we'd make use of it.